A/B Tests: Propagate backend-loaded A/B assignments into client

jsnmoon opened the pull request:
Fixes #20097.

This change, in conjunction with `D9429-code`, enables users to fetch their assigned A/B test variations stored as user attributes from wpcom REST API. Specifically, this is done by adding A/B variation data into the API request to `wpcom.me().get( ... )`.


#### Testing instructions

0. Apply `D9429-code` to your sandbox and sandbox `public-api.wordpress.com`.
1. Check out locally (`fix/load-ab-variations-into-backend`) and start your server.
2. Set your localStorage debug flag to `calypso:user` like so:  
```javascript
localStorage.debug = 'calypso:user';
```
3. Open the [homepage](http://calypso.localhost:3000/).
4. Verify that you see the following debug lines:
```
calypso:user Initializing User
calypso:user Getting user from api
calypso:user User successfully retrieved
```
5. Find the corresponding network request for variation assignments, which should look like this:

![me_request_with_a_b_variations](https://user-images.githubusercontent.com/4044428/35293933-7ca21cee-0032-11e8-8668-934e8e58bbd3.png)

6. Ensure that localStorage has been populated with expected values. With my user, it looks like the following:

``` javascript
> localStorage.ABTests
"{"abtest_skipThemesSelectionModal_20170904":"show"}"

> JSON.parse(localStorage.wpcom_user).abtests
{ abtest_skipThemesSelectionModal_20170904: "show" }
```

#### Reviews

- [x] Code (ensure that the code changes are reasonable)
- [x] Product (ensure that the product works as intended)

yoavf commented on client/lib/user/shared-utils.js:
```diff
@@ -75,8 +74,8 @@ export default {
-		var language = getLanguage( attributes.language ),
-			primayBlogUrl = attributes.primary_blog_url || '';
+		const language = getLanguage( attributes.language );
+		const primayBlogUrl = attributes.primary_blog_url || '';
```
Good occasion to fix the typo here (primary/primay)

jsnmoon replied:
👍

yoavf commented on client/lib/user/user.js:
```diff
@@ -67,31 +68,42 @@ User.prototype.initialize = function() {
 		if ( this.data ) {
-			this.clearStoreIfChanged( this.data.ID );
-			store.set( 'wpcom_user', this.data );
+			if ( ! this.data.abtests ) {
```
Should this part be in this PR? seems like it's only for the bootstrapped user load

jsnmoon replied:
With the latest change loading variations from `/me`, this should no longer be an issue.

yoavf commented:
@jsnmoon thanks for taking this on!

It works as described , but I'm wondering if this even makes sense to do outside of user bootsrapping - in my test it was rather slow, the `me/abtests` request queued up after a few other requests, and taking quite a while to load. 

Have you considered fetching the abtest data as part of the initial `/me` request? I'm not sure if the additional data input/output makes sense.

jsnmoon commented:
> It works as described, but I'm wondering if this even makes sense to do outside of user bootstrapping - in my test it was rather slow, the me/abtests request queued up after a few other requests, and taking quite a while to load.
> 
> Have you considered fetching the abtest data as part of the initial /me request? I'm not sure if the additional data input/output makes sense.

Great idea! I'll update this PR (as well as the backend one) to emit A/B variation data through the initial `/me` request instead.

jsnmoon commented:
Rebased from `master`.

jsnmoon commented:
Rebased from master, refactored `'ABTests'` LocalStorage key as an exported `const`, and added a parameter to `getActiveTestNames` for controlling datestamp appending.

---

Any chance we could get a 👍 / 👎 / 🚢 ? It would be great to push this out this week or the next.

samouri commented on client/lib/user/user.js:
```diff
@@ -212,14 +232,8 @@ User.prototype.getAvatarUrl = function( options ) {
-	}
-
-	return isRTL;
+	const language = this.getLanguage();
```
nice refactor 👍

samouri commented on client/my-sites/stats/checklist-banner/index.jsx:
```diff
@@ -110,7 +111,7 @@ export class ChecklistBanner extends Component {
 		}
 
-		const abtests = store.get( 'ABTests' );
+		const abtests = store.get( ABTEST_LOCALSTORAGE_KEY );
```
I know this isn't the main thrust of your PR, but this file definitely seems like it shouldn't be relying directly on `store`. Ideally it would rely on an abtest lib

jsnmoon replied:
Definitely agreed, this is an anti-pattern. I'll add a comment noting it as such, but it'd be even worse if we had just left `'ABTests'` string in there untouched.

samouri commented on client/lib/user/user.js:
```diff
@@ -69,6 +70,9 @@ User.prototype.initialize = function() {
 		if ( this.data ) {
 			this.clearStoreIfChanged( this.data.ID );
 			store.set( 'wpcom_user', this.data );
```
its kind of strange that `wpcom_user` will also by default have all of the abtests instead of just the abtests key.
Do you see abtests as being a sensical part of the user object or do you think it just makes sense to piggyback on its api call + then remove it from the object before setting to localstorage?

jsnmoon replied:
I'm not sure if I follow your reasoning - this block is ingesting data from user bootstrapping and making it available in local storage. It will only include the user's assigned A/B test variations instead of `all of the abtests`. (For example, my user object currently contains only one key-value pair for our running A/B tests.)

I believe having the user's assigned A/B variations in the user object as appropriate.

samouri replied:
so i guess my point is this.  Why store it in two places when one will do?

```jsx
const getABTests = () => store.get( 'wpcom_user ).abtests;
```

samouri commented on client/lib/user/user.js:
```diff
@@ -69,6 +70,9 @@ User.prototype.initialize = function() {
 		if ( this.data ) {
 			this.clearStoreIfChanged( this.data.ID );
 			store.set( 'wpcom_user', this.data );
+			if ( this.data.abtests ) {
```
the code in this section looks just about identical to the `handleSuccess` for the user fetch that you've made later in the code.
Are there any differences in the two and if not can we reuse the fn?

jsnmoon replied:
The only difference seems to be the handling of settings in the `if ( this.settings ) { ... }` block. I'll see if I can refactor this further.

jsnmoon commented on client/lib/user/user.js:
```diff
@@ -120,62 +124,78 @@ User.prototype.fetch = function() {
 	this.fetching = true;
 	debug( 'Getting user from api' );
 
 	me.get(
```
Unfortunately, I don't think [that's the case](https://github.com/Automattic/wpcom.js/blob/master/lib/me.js#L24) ([request method here](https://github.com/Automattic/wpcom.js/blob/ab065dfc1e515aba6e36cb9b6b438366dff9bc3c/lib/util/request.js#L14)).

samouri replied:
Follow the rabbit hole one level deeper and you will find [sendRequest](https://github.com/Automattic/wpcom.js/blob/ab065dfc1e515aba6e36cb9b6b438366dff9bc3c/lib/util/send-request.js#L88) that returns a promise :)

jsnmoon commented:
This was reverted in #22147 due to causing outages on `/log-in`, `/start`, and `/start/account`. Luckily, our e2e canary tests caught the problem before it could be released to production.
